cpp/src/arrow/util/iterator.h

```diff
@@ -82,4 +83,38 @@ std::unique_ptr<VectorIterator<T>> MakeIterator(std::vector<T> v) {
   return std::unique_ptr<VectorIterator<T>>(new VectorIterator<T>(std::move(v)));
 }
 
+template <typename T>
+class EmptyIterator : public Iterator<T> {
+ public:
+  explicit EmptyIterator(Status s = Status::OK()) : status_(s) {}
+
+  Status Next(T* out) override {
+    *out = NULLPTR;
+    return status_;
+  }
+
+ private:
+  Status status_;
+};
+
+template <typename Fn, typename T>
+class FunctionIterator : public Iterator<T> {
+ public:
+  using IteratorType = Iterator<T>;
+
+  explicit FunctionIterator(Fn fn) : fn_(std::move(fn)) {}
+
+  Status Next(T* out) override { return fn_(out); }
+
+ private:
+  Fn fn_;
+};
+
+template <typename Fn, typename T = typename std::remove_pointer<
+                           single_call::argument_type<0, Fn>>::type>
+std::unique_ptr<FunctionIterator<Fn, T>> MakeIterator(Fn fn) {
```

Comment:
Call this `MakeFunctionIterator`?

Reply:
because `MakeIterator` was shorter. I can change it